static/js/services/delete-docs.js

```diff
@@ -60,13 +62,44 @@ var _ = require('underscore');
         });
       };
 
+      var bulkDeleteRemoteDocs = function (docs, eventListeners) {
+        var deferred = $q.defer();
+        var xhr = new XMLHttpRequest();
+        xhr.onprogress = function() {
+          if (xhr.responseText) {
+            var currentResponse = partialParse(xhr.responseText);
+            var totalDocsDeleted = _.flatten(currentResponse).length;
+            if (eventListeners.progress && Array.isArray(currentResponse)) {
+              eventListeners.progress(totalDocsDeleted);
+            }
+          }
+        };
+        xhr.onload = function() {
+          if (this.status >= 200 && this.status < 300) {
+            deferred.resolve(_.flatten(JSON.parse(xhr.response)));
+          } else {
+            deferred.reject(new Error('Server responded with ' + this.status + ': ' + xhr.statusText));
+          }
+        };
+        xhr.onerror = function() {
+          deferred.reject(new Error('Server responded with ' + this.status + ': ' + xhr.statusText));
+        };
+        xhr.open('POST', '/api/v1/bulk-delete', true);
+        xhr.setRequestHeader('Content-type', 'application/json');
+        xhr.send(JSON.stringify({ docs: docs }));
+        return deferred.promise;
+      };
+
       /**
        * Delete the given docs. If 'person' type then also fix the
        * contact hierarchy.
```

Comment:
Can you go into more detail about "fixing" the contact hierarchy?

It's not clear to me from reading the code why we're getting parents and then writing them without seemingly changing their values at all?

Reply:
So I didn't write this comment or the "fixing" code, but after working with it for a bit it appears that "fixing" the hierarchy just means if the document to be bulk deleted is the child of another doc and also its `contact`, we should update the `contact` field on the parent to `null`. Happy to add this in to the comment for future code wanderers

Reply:
Yeah, an explanation would be helpful for sure! Also, while you're there, can you rename the `getParent` function to something more appropriate? I literally didn't understand this because I didn't read what that function did, figuring it would get parents. As opposed to get all parents, filter them for ones where the parent's contact is the child and also delete the contact 😱